provisioner/claim-reconciler/reconiler.go

```diff
@@ -0,0 +1,113 @@
+package object_bucket_claim_reconciler
+
+import (
+	"context"
+	"fmt"
+	"github.com/yard-turkey/lib-bucket-provisioner/pkg/apis/objectbucket.io/v1alpha1"
+	"github.com/yard-turkey/lib-bucket-provisioner/provisioner"
+	"k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"strings"
+	"time"
+)
+
+type objectBucketClaimReconciler struct {
+	client          client.Client
+	provisionerName string
+	provisioner     provisioner.Provisioner
+	retryInterval   time.Duration
+	retryTimeout    time.Duration
+	retryBackoff    int
+}
+
+var _ reconcile.Reconciler = &objectBucketClaimReconciler{}
+
+type ReconcilerOptions struct {
+	RetryInterval time.Duration
+	RetryTimeout  time.Duration
+	RetryBackoff  int
+}
+
+func NewObjectBucketClaimReconciler(c client.Client, name string, provisioner provisioner.Provisioner, options ReconcilerOptions) *objectBucketClaimReconciler {
+	return &objectBucketClaimReconciler{
+		client:          c,
+		provisionerName: strings.ToLower(name),
+		provisioner:     provisioner,
+		retryInterval:   options.RetryInterval,
+		retryTimeout:    options.RetryTimeout,
+		retryBackoff:    options.RetryBackoff,
+	}
+}
+
+// TODO this is the guts of our controller.
+//   `request` is a 'namespace/name' of a resource.
+func (r *objectBucketClaimReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// /   ///   ///   ///   ///   ///   ///
+	// TODO    CAUTION! UNDER CONSTRUCTION!
+	// /   ///   ///   ///   ///   ///   ///
+	// This process should basically
+	// 1. get the claim resource
+	// 2. get the class for the claim
+	// 3. verify the class.provisioner matches this provisioner name
+	// 4. attempt to provision the bucket
+	// 5. create the OB returned from Provision()
+	// 6. create the secret and config map in the OBC namespace
+
+	obc := &v1alpha1.ObjectBucketClaim{}
+
+	if err := r.client.Get(context.TODO(), request.NamespacedName, obc); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	claimClass := obc.Spec.StorageClass
+
+	if !r.shouldProvision(claimClass) {
+		return reconcile.Result{}, fmt.Errorf("Unrecognized storageClass.provisioner name")
+	}
+
+	// TODO wrap in retry logic
+	ob, s3keys, err := r.provisioner.Provision(obc)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if s3keys.AreEmpty() {
+		msg := "s3 access key or secret key is nil"
+		err = r.provisioner.Delete(obc)
```

Comment:
Is this block saying "Provision() created a bkt without using (or returning) the s3 key-pairs, so therefore we need to delete that physical bkt now"?
What if (down the road) reclaimPolicy is "retain"? I know this is WIP but just sharing some initial thoughts...

Reply:
RE: retainPolicy - see my comment [here](https://github.com/rook/rook/pull/2625/commits/4a94959fea87008281fafcd4cca950fb80351b66#r260813528).  From what I can tell, k8s has obsolesced retainPolicy in the external storage library.